internal/impl/crdb/input_changefeed.go

```diff
@@ -0,0 +1,245 @@
+package crdb
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/Jeffail/benthos/v3/internal/shutdown"
+	"github.com/Jeffail/benthos/v3/public/service"
+	"github.com/jackc/pgx/v4"
+	"github.com/jackc/pgx/v4/pgxpool"
+)
+
+var (
+	sampleString = `{
+	"primary_key": "[\"1a7ff641-3e3b-47ee-94fe-a0cadb56cd8f\", 2]", // stringifed JSON array
+	"row": "{\"after\": {\"k\": \"1a7ff641-3e3b-47ee-94fe-a0cadb56cd8f\", \"v\": 2}, \"updated\": \"1637953249519902405.0000000000\"}", // stringified JSON object
+	"table": "strm_2"
+}`
+)
+
+func crdbChangefeedInputConfig() *service.ConfigSpec {
+	return service.NewConfigSpec().
+		Categories("Integration").
+		Summary(fmt.Sprintf("Listens to a CockroachDB Core Changefeed and creates a message for each row received. Each message is a json object looking like: \n```json\n%s\n```", sampleString)).
+		Description("Will continue to listen to the Changefeed until shutdown. If provided, will maintain the previous `timestamp` so that in the event of a restart the chanfeed will resume from where it last processed. This is at-least-once processing, as there is a chance that a timestamp is not successfully stored after being emitted and therefore a row may be sent again on restart.\n\nNote: You must have `SET CLUSTER SETTING kv.rangefeed.enabled = true;` on your CRDB cluster").
+		Field(service.NewStringField("dsn").
+			Description(`A Data Source Name to identify the target database.`).
+			Example("postgresql://dan:xxxx@free-tier.gcp-us-central1.cockroachlabs.cloud:26257/defaultdb?sslmode=require&options=--cluster%3Dportly-impala-2852")).
+		Field(service.NewStringField("root_ca").
+			Description(`A Root CA Certificate`).
+			Example(`-----BEGIN CERTIFICATE-----
+FEIWJFOIEW...
+-----END CERTIFICATE-----`).
+			Optional()).
+		Field(service.NewStringListField("tables").
+			Description("CSV of tables to be included in the changefeed").
+			Example([]string{"table1", "table2"})).
+		Field(service.NewStringListField("options").
+			Description("CSV of options to be included in the changefeed (WITH X, Y...).\n**NOTE: The CURSOR option here will only be applied if a cached cursor value is not found. If the cursor is cached, it will overwrite this option.**").
+			Example([]string{"UPDATE", "CURSOR=1536242855577149065.0000000000"}).
+			Optional())
+}
+
+type crdbChangefeedInput struct {
+	dsn        string
+	ctx        context.Context
+	cancelFunc context.CancelFunc
+	pgConfig   *pgxpool.Config
+	pgPool     *pgxpool.Pool
+	rootCA     string
+	statement  string
+
+	rows pgx.Rows
+
+	tables []string
+	optins []string
+
+	logger  *service.Logger
+	shutSig *shutdown.Signaller
+}
+
+func newCRDBChangefeedInputFromConfig(conf *service.ParsedConfig, logger *service.Logger) (*crdbChangefeedInput, error) {
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	c := &crdbChangefeedInput{
+		logger:     logger,
+		shutSig:    shutdown.NewSignaller(),
```

Comment:
Thanks for this work! I'm trying to understand the signalling logic and I was wondering if this `shutSig` was meant to be used somehow in `crdbChangefeedInput.Read()`. Also, instead of having a cancellable context as a struct member, I think a quit chan might be more idiomatic: https://go.dev/blog/context-and-structs

Reply:
`shutSig` was something that I pulled from other inputs (as examples). I had assumed it was needed at a higher level. I can remove if it's not required from an input as the contexts handle the shutting down. Shall I remove?

For the cancelleable context, that is needed to kill the blocking query found in the `Read()` function. Without it we cannot gracefully exit the `c.pgPool.Query()` (blocks until first row is available) or `c.rows.Next()` (blocks until cancelled).

Unless I am missing something here. Still _relatively_ new to go semantics.

Reply:
Sure, no worries! I'm no expert either, but @Jeffail might be able to offer more insight into what would be a good way to make use of this `shutSig` in this input, if it makes sense to do so. Regarding the context part, yep, you can use it like you did, but the link I pasted above offers details about why one should avoid using contexts that way. If all you need is to signal from one goroutine to another to exit, you can simply `make` a channel of `struct{}`. If you close it in one goroutine, the other goroutine that's waiting to read from it will immediately unblock. See the ["channel axioms"](https://dave.cheney.net/2014/03/19/channel-axioms) for details.

Reply:
I found that the shutdonw signal was not needed in my testing, so I have removed it. I can add back in if advised but it did not seem to have any impact on operation.

Reply:
I guess my issue with the chan vs. context is how the query perceives it. I am assuming since the Read() context will be Done() when SIGTERM signal is sent, then the query should gracefully exit anyway. My worry was that the goroutine would still be consuming new rows while the Read()ing stopped, but I see this not to be the case...

I removed the context in the struct and use the function based on instead. Seems to have no impact on operations either. Thanks for the advice!

Reply:
So we definitely need a context for that Query that is NOT shared with the Read(), seems like we need to keep it in the struct unfortunately due to the nature that the Read() context is cancelled after the read finishes.

Reply:
I have found that after a single read, no more reads are parsed when using the Read() context

Reply:
If we don't store this context somewhere we can cancel it (and not the same as the Read() context), then the Read() will continue to block and we will hit the timeout. The context needs to be cancelled and if we are in anything after the first read run we will no longer have access to the same context. I think this is a case where we have to put it in the struct due to the nature of the Read() context being cancelled after it is first called.

Reply:
@mihaitodor Do you have any other suggestions? Otherwise I cannot access the context in the query and run the risk of messages getting consumed but dropped

Reply:
I'll defer to Ash for further feedback. I don't understand this well enough to offer proper guidance... Sorry about that!

Reply:
@Jeffail Just wanted to send a friendly ping on whether you have any feedback here :)

Definitely don't think I can get away with not cancelling the context since that goroutine will block forever.

Reply:
@mihaitodor One thing I've been thinking about is how we might be able to make compromises. My concern around not cancelling the query context has mostly been due to reading rows unnecessarily. I think this might be mitigated by 2 means, but I definitely need some confirmation:

1. benthos has already stopped reading, so returning new rows doesn't actually cause anything to happen
2. My other worry was when I v2 this intgeration and add in cursor caching. As long as the shutdown signal cuts of the cursor caching immediately, we don't need to worry about any subsequent hidden reads that go on in the background updating the cursor without emitting messages
3. Since benthos is shutting down it's easy enough for my to just kill the pool connection, that should shut things down similar to cancelling a query.

I am currently implementing these changes locally to see how they work.

Reply:
*Edited the wording to be more accurate

Reply:
I've just made the changes that remove the stored context, if we abort the shutdown and let the ending of the process terminate the DB connections (.Close() on row or pool will block due to the query) then everything works without stored context OR a shutdown channel.

Reply:
Just checking in on this again :)

Reply:
@danthegoodman1, the core mechanisms here look fine to me, the call to `Close` just needs to close any currently open `pgPool` and `rows`. Ideally we need some form of integration tests next as otherwise I can't really help out from my end.

Reply:
@Jeffail I can only do that if I am able to cancel the long running query with a context, which means I need to keep a context in the struct that I can cancel across different Read() calls. This was something that was mentioned to avoid, but I can add that back in.

another way to handle that would be another goroutine but that would definitely involve more code. Would keep the context out of the struct though so I can try that way first.

If I can’t cancel that query then the close on rows and pgpool will hang.

Reply:
@Jeffail @mihaitodor using nested goroutines I was able to do query cancellation without storing a context in the struct. Let me know if that works as an alternative to storing the context in the struct.

I will get an integration test working, will probably involve needing to run a CRDB container locally but otherwise should be quite simple. Will try to do that Sunday :)

Reply:
Just want to note that I haven't forgotten about this, just have been super busy with work lately. Hoping to hit this by the end of the weekend.

Reply:
Yeah, I think that addresses my comment, thanks!